addons/mail/static/src/models/thread_cache/thread_cache.js

```diff
@@ -94,18 +94,88 @@ function factory(dependencies) {
         // Private
         //----------------------------------------------------------------------
 
+        /**
+         * @private
+         * @returns {mail.message[]}
+         */
+        _computeFetchedMessages() {
+            if (!this.thread) {
+                return [['unlink-all']];
+            }
+            const toUnlinkMessages = [];
+            for (const message of this.fetchedMessages) {
+                if (!this.thread.messages.includes(message)) {
+                    toUnlinkMessages.push(message);
+                }
+            }
+            return [['unlink', toUnlinkMessages]];
+        }
```

Comment:
What is the reasoning for removing from `fetchedMessages` those that are not in `messages`?

Feels like `fetchedMessages` should only ever be written by the server conceptually.

Then if we need a "clean" variable, it should be computed separately?

Reply:
The way messages are fetched is simplistic and this explains the holes in message list issue. The workaround to not have holes consists of having `fetchedMessages` that contains list of messages received from `message_fetch`, so that they are in sequence.

The reason we need to adjust is to avoid considering "fetched messages" that are no longer in thread, e.g. by being deleted. If we create another field for "clean" fetched messages, this adds more complexity with lots of fields with obscure naming (e.g. "serverFetchedMessages"... "server" and "fetched" look redundant or "fetchedInThreadMessages" which leads to more obscure naming for fields that are more meaningful in business code...).

The compromise is to not introduce another field and simply adjust "fetchedMessages" like here, as this keeps things under control (easy to understand + works with simplistic server logic).